### PR TITLE
Migrate entity-related codegen to ProtoData

### PR DESCRIPTION
**Pull request by alexander-yevsyukov**
This PR initiates the migration of codegen processing to ProtoData. The main part of this PR is the migration of codegen of `Column` classes in messages that implement the `EntityState` interface.

### In details

`EntityPlugin` was introduced as an umbrella for all codgen operations related to `EntityState` messages.

The plugin is based on `EntityDiscoveryProcess`, which transforms `TypeDiscovered` event to `EntityStateDiscovered` event, if the discovered message type has appropriate options. Other parts of the `EntityPlugin` would listen to the event and act accordingly.

The first render of the plugin is `ColumnRenderer` which renders a `public static` class `Column` nested under corresponding message type implementing `EntityState`.

`ColumnRenderer` and `ColumnClassFactory` now do what `ColumnGen` class used to do. The `ColumnGen` class was deprecated and is no longer added to `protoc` plugin of McJava.

### Other notable changes
 * `CodegenOptionsConfig ` class was renamed to `MessageCodegenOptions `.
 * `MessageGroupConfig` class was renamed to `ConfigWithFields `.
 * File name pattern patterns are no longer supported for signifying custom types of messages. Please use mixin interfaces or custom code generation.
 * Validation runtime dependency is added to a project to which McJava applied anyway, because generated code implements `ValidatableMessage`  even if Validation codegen is turned off, and we want new run-time validation being available anyway.
 * Extensions recently moved to `base` and `tool-base` were applied.
 * Improved filtering for the `mc-java-plugin-bundle` artifact.

### CODE REVIEW

**Comment by alexander-yevsyukov** [comment]
@armiol, just FYI. It finally builds again. 

“Getting better, every day...” :)

File: buildSrc/src/main/kotlin/module.gradle.kts [hunk 1]
```diff
@@ -137,6 +138,10 @@ fun Module.forceConfigurations() {
                     KotlinX.Coroutines.jdk8,
                     Protobuf.compiler,
                     Grpc.api,
+                    Grpc.core,
+                    Grpc.protobuf,
+                    Grpc.stub,
+                    "org.yaml:snakeyaml:2.1",
```
Comment (armiol) [inline buildSrc/src/main/kotlin/module.gradle.kts:144]: Please extract to the corresponding dependency object.

File: mc-java-base/src/main/proto/spine/tools/mc/java/config.proto [hunk 2]
```diff
@@ -98,12 +98,16 @@ message Signals {
 
     // The patterns which define the group.
     //
-    // It is recommended that Spine users define events in files ending with "events.proto",
-    // rejections — with "rejections.proto", and commands — with "commands.proto". However, users
-    // can override this convention by changing these patterns.
+    // It is recommended that users define signal message types according
+    // to the following conventions:
+    //  1. commands — in files ending with "commands.proto",
+    //  3. events — in files ending with "events.proto", and
```
Comment (armiol) [inline mc-java-base/src/main/proto/spine/tools/mc/java/config.proto:104]: This should go under 2. 
And the next one should go under 3.

Or please explain the idea.

File: mc-java-entity/src/main/kotlin/io/spine/tools/mc/entity/EntityDiscoveryProcess.kt [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+import io.spine.server.entity.alter
+import io.spine.server.event.React
+import io.spine.server.procman.ProcessManager
+import io.spine.tools.mc.entity.event.EntityStateDiscovered
+import io.spine.tools.mc.entity.event.entityStateDiscovered
+import io.spine.tools.mc.java.codegen.Entities
+import java.util.Optional
+
+/**
+ * Gathers entity states declared in proto files.
```
Comment (armiol) [inline mc-java-entity/src/main/kotlin/io/spine/tools/mc/entity/EntityDiscoveryProcess.kt:42]: Ideally, the process should just transform the events by filtering out those not matching our Entity conventions. But not gather. Gathering is a responsibility of some projection, which in this case would be subscribed to `EntityStateDiscovered`.

Let's discuss.

File: mc-java-entity/src/main/kotlin/io/spine/tools/mc/java/entity/EntityStateRenderer.kt [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        if (!relevant) {
+            return
+        }
+        val entityStates = foundEntityStates()
+        entityStates.forEach {
+            doRender(it, sources)
+        }
+    }
+
+    abstract fun doRender(type: MessageType, sources: SourceFileSet)
```
Comment (armiol) [inline mc-java-entity/src/main/kotlin/io/spine/tools/mc/java/entity/EntityStateRenderer.kt:62]: Please document.

File: mc-java-entity/src/main/kotlin/io/spine/tools/mc/java/entity/column/ColumnAccessor.kt [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+import io.spine.tools.psi.java.Environment.elementFactory
+import org.intellij.lang.annotations.Language
+
+/**
+ * Generates a method which returns a [strongly typed][EntityColumn] entity column.
+ *
+ * The name of the method matches the name of the [entity state][io.spine.base.EntityState]
+ * converted to [javaCase].
+ */
+internal class ColumnAccessor(
```
Comment (armiol) [inline mc-java-entity/src/main/kotlin/io/spine/tools/mc/java/entity/column/ColumnAccessor.kt:49]: I would name it `ColumnAccessorMethod`. Or `ColumnAccessorGen`.

Right now, located next to `ColumnRenderer`, it looks as something that reads column values. While it does not do that, it just generated a code which does.

File: mc-java-entity/src/main/kotlin/io/spine/tools/mc/java/entity/column/ColumnClassFactory.kt [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+import io.spine.tools.psi.java.createPrivateConstructor
+import io.spine.tools.psi.java.makeFinal
+import io.spine.tools.psi.java.makePublic
+import io.spine.tools.psi.java.makeStatic
+import io.spine.tools.psi.java.topLevelClass
+import java.lang.String.format
+import org.intellij.lang.annotations.Language
+
+/**
+ * Creates a class called `Column` and nest it under the top level entity state class.
```
Comment (armiol) [inline mc-java-entity/src/main/kotlin/io/spine/tools/mc/java/entity/column/ColumnClassFactory.kt:56]: "nests".

File: mc-java-protoc/src/main/java/io/spine/tools/mc/java/protoc/Plugin.java [hunk 7]
```diff
@@ -81,7 +80,10 @@ public static void main(String[] args) {
                 MethodGen.instance(config),
                 new BuilderGen(),
                 NestedClassGen.instance(config),
-                ColumnGen.instance(config),
+
+                // Use new rendering of columns.
+                //ColumnGen.instance(config),
```
Comment (armiol) [inline mc-java-protoc/src/main/java/io/spine/tools/mc/java/protoc/Plugin.java:85]: Not sure why this is commented out. Let's discuss.

**Review by armiol — CHANGES_REQUESTED** [review]
@alexander-yevsyukov please see my comments.

Other that that, great PR.

**Review by armiol — APPROVED** [review]

**Pull request merged**